File: WsServer/WsServer.Shared/Rooms/Room.cs
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;

namespace WsServer.Rooms;

/// <summary>
/// Represents a room with specific communication modes and client membership
/// </summary>
public class Room
{
    public string Id { get; }
    public string Name { get; set; }
    public HashSet<CommunicationMode> SupportedModes { get; }
    public ConcurrentDictionary<uint, RoomClient> Clients { get; }
    public DateTime CreatedAt { get; }
    public uint? HostClientId { get; set; }
    
    public int ClientCount => Clients.Count;
    
    public Room(string id, string name, IEnumerable<CommunicationMode> supportedModes, bool isPersistent = false)
    {
        Id = id;
        Name = name;
        SupportedModes = new HashSet<CommunicationMode>(supportedModes);
        Clients = new ConcurrentDictionary<uint, RoomClient>();
        CreatedAt = DateTime.UtcNow;
        IsPersistent = isPersistent;
    }
    
    /// <summary>
    /// Whether this room should be kept even when empty (e.g., default rooms like lobby)
    /// </summary>
    public bool IsPersistent { get; }
    
    /// <summary>
    /// Check if the room supports a specific communication mode
    /// </summary>
    public bool SupportsMode(CommunicationMode mode)
    {
        return SupportedModes.Contains(mode);
    }
    
    /// <summary>
    /// Add a client to the room
    /// </summary>
    public bool AddClient(uint clientId, string clientName)
    {
        return Clients.TryAdd(clientId, new RoomClient(clientId, clientName));
    }
    
    /// <summary>
    /// Remove a client from the room
    /// </summary>
    public bool RemoveClient(uint clientId)
    {
        return Clients.TryRemove(clientId, out _);
    }
    
    /// <summary>
    /// Get client by ID
    /// </summary>
    public RoomClient? GetClient(uint clientId)
    {
        return Clients.GetValueOrDefault(clientId);
    }
    
    /// <summary>
    /// Get all client IDs in the room
    /// </summary>
    public IEnumerable<uint> GetClientIds()
    {
        return Clients.Keys;
    }
    
    /// <summary>
    /// Check if room has any clients
    /// </summary>
    public bool IsEmpty => Clients.IsEmpty;
}

/// <summary>
/// Represents a client within a room
/// </summary>
public class RoomClient
{
    public uint Id { get; }
    public string Name { get; }
    public DateTime JoinedAt { get; }
    public CommunicationMode CurrentMode { get; set; }
    
    public RoomClient(uint id, string name)
    {
        Id = id;
        Name = name;
        JoinedAt = DateTime.UtcNow;
        CurrentMode = CommunicationMode.TextChat; // Default mode
    }
}

File: WsCore.Client/src/network/rooms/Room.ts
import { CommunicationMode } from './CommunicationMode.js';
import { RoomClient } from './RoomClient.js';

/**
 * Represents a room with specific communication modes and client membership
 */
export class Room {
  public readonly id: string;
  public name: string;
  public readonly supportedModes: Set<CommunicationMode>;
  public readonly clients: Map<number, RoomClient>;
  public readonly createdAt: Date;
  public hostClientId?: number;
  public userCount: number = 0;
  public isPersistent: boolean = false; // Default rooms should not be deleted

  constructor(
    id: string,
    name: string,
    supportedModes: CommunicationMode[],
    isPersistent: boolean = false
  ) {
    this.id = id;
    this.name = name;
    this.supportedModes = new Set(supportedModes);
    this.clients = new Map();
    this.createdAt = new Date();
    this.isPersistent = isPersistent;
  }

  /**
   * Check if the room supports a specific communication mode
   */
  public supportsMode(mode: CommunicationMode): boolean {
    return this.supportedModes.has(mode);
  }

  /**
   * Add a client to the room
   */
  public addClient(clientId: number, clientName: string, mode?: CommunicationMode): boolean {
    if (!this.clients.has(clientId)) {
      const client = new RoomClient(clientId, clientName);
      if (mode && this.supportsMode(mode)) {
        client.currentMode = mode;
      }
      this.clients.set(clientId, client);
      return true;
    }
    return false;
  }

  /**
   * Remove a client from the room
   */
  public removeClient(clientId: number): boolean {
    return this.clients.delete(clientId);
  }

  /**
   * Get client by ID
   */
  public getClient(clientId: number): RoomClient | undefined {
    return this.clients.get(clientId);
  }

  /**
   * Get all client IDs in the room
   */
  public getClientIds(): number[] {
    return Array.from(this.clients.keys());
  }

  /**
   * Check if room has any clients
   */
  public get isEmpty(): boolean {
    return this.clients.size === 0;
  }

  /**
   * Get client count
   */
  public get clientCount(): number {
    return this.clients.size;
  }
}
